### ISSUE TITLE
Lets Make a List of ToDos and RoadMap file

### ISSUE THREAD

**Issue by bawarkamalqader** [opentypejs/opentype.js#374]
Just wanted to ask if it was possible to make a ToDo list , so that new contributors will be able to know where to start, and what to do , like a ROADMAP.

things like ::
- GPOS Table Writing
- GSUB Subtables Writing : 2,5,6,7,8,9
- others ...

**Comment by brawer** [comment]
Fix [test failures](https://rawgit.com/unicode-org/text-rendering-tests/master/reports/OpenType.js.html)?

**Comment by bawarkamalqader** [comment]
that goes into implementing remaining tables, currently 18 tables are supported, we need to add the remaining tables ::

Remaining Tables  :
 'acnt' , 'ankr' , 'avar' , 'bdat' , 'bhed' , 'bloc' , 'bsln' , 'cvar' , 'cvt ' , 'EBSC' , 'fdsc' , 'feat' , 'fmtx' , 'fond' , 'fpgm' , 'gasp' , 'gcid' , 'gvar' , 'hdmx' , 'just' , 'kerx' , 'lcar' , 'mort' , 'morx' , 'opbd' , 'prep' , 'prop' , 'sbix' , 'trak' , 'vhea' , 'vmtx' , 'xref' , 'Zapf' , 

Important Opentype Fonts
 BASE,CBDT,CBLC,COLR,CPAL,DSIG,EBDT,EBLC,GDEF,HVAR,JSTF,LTSH,MATH,MERG

**Comment by bawarkamalqader** [comment]
we need to sort the remaining tables by priority , i mean the tables that need to be supported earlier and implemented immediately .

**Comment by brawer** [comment]
I'm in no position to define a roadmap, but personally I'd recommend fixing the security bugs first; then fixing the missing parts of cmap; then fixing shaping (GSUB/GPOS). The latter will need GDEF (easy), various barely-documented script-specific shaping engines (painful) plus the Universal Shaping Engine USE. The fontkit project has ported HarfBuzz to JavaScript, so have a look at the fontkit sources. Once all that is done, OpenType.js will be able to correctly render Unicode text.

Personally, I would recommend looking at variable fonts (gvar, hvar, avar, HVAR), color fonts, AAT tables, etc. only after the basics of text rendering are working without bugs. Just my personal two cents of course; others might have different opinions.

**Comment by Jolg42** [comment]
Having a list sorted by priority would be great indeed! Thanks @brawer for your input.

@bawarkamalqader are you interested in something specific?

Adding (and first finding) good fonts with the associated tables to the list would be great as well to make sure everything works well.

About cmap adding the missing parts will probably need some refactoring but not for GSUB. So starting there will be easier. Also I recommend to start on the implementing reading first  and then writing, both with tests.

**Comment by solomancode** [comment]
I was able to compose an overview of the current library status, **what's been done** and **what needs to be done**. each item has a severity value to indicate the importance of this item. I slightly adjusted Jakob's Nielsen [Severity Scale](<https://www.nngroup.com/articles/how-to-rate-the-severity-of-usability-problems/>) to match the context of this document.



**<u>Severity Scale</u>**

**0** = I don't agree that this is a usability problem at all
**1** = Need not be fixed unless extra time is available
**2** = Minor usability problem: fixing this should be given low priority
**3** = Major usability problem: important to fix, so should be given high priority
**4** = Usability catastrophe: imperative to fix this before product can be released





**<u>Required Tables</u>**

| description                            | status |
| -------------------------------------- | ------ |
| **cmap**	character to glyph mapping | ✔️      |
| **head**	font header                | ✔️      |
| **hhea**	horizontal header          | ✔️      |
| **hmtx**	horizontal metrics         | ✔️      |
| **loca**	index to location          | ✔️      |
| **maxp**	maximum profile            | ✔️      |
| **name**	naming                     | ✔️      |
| **post**	PostScript                 | ✔️      |

---

**<u>Optional Tables</u>**

| description                              | status    | severity | reference                                                    |
| ---------------------------------------- | --------- | -------- | ------------------------------------------------------------ |
| **cvt** 	control value                | *missing* | **2**    | [reference](https://docs.microsoft.com/en-us/typography/opentype/spec/cvt) |
| **fpgm**	font program                 | *missing* | **2**    | [reference](https://docs.microsoft.com/en-us/typography/opentype/spec/fpgm) |
| **hdmx**	horizontal device metrics    | *missing* | **2**    | [reference](https://docs.microsoft.com/en-us/typography/opentype/spec/hdmx) |
| **kern**	kerning                      | ✔️         | **4**    |                                                              |
| **OS/2**	OS/2                         | ✔️         | **3**   |                                                              |
| **prep**	control value program        | *missing* | **2**   | [reference](<https://docs.microsoft.com/en-us/typography/opentype/spec/prep>) |
| **VORG**	Vertical Origin              | *missing* | **2**   | [reference](<https://docs.microsoft.com/en-us/typography/opentype/spec/vorg>) |
| **gasp**	Grid-fitting/Scan-conversion | *missing* | **2**   | [reference](<https://docs.microsoft.com/en-us/typography/opentype/spec/gasp#39gasp39-table-formats>) |
| **DSIG**	Digital signature |*missing*|**1**|[reference](https://docs.microsoft.com/en-us/typography/opentype/spec/dsig)|
| **LTSH**	Linear threshold data |*missing*|**2**|[reference](https://docs.microsoft.com/en-us/typography/opentype/spec/ltsh)|
| **MERG**	Merge |*missing*|**2**|[reference](https://docs.microsoft.com/en-us/typography/opentype/spec/merg)|
| **meta**	Metadata |*missing*|**1**|[reference](https://docs.microsoft.com/en-us/typography/opentype/spec/meta)|
| **STAT**	Style attributes |*missing*|**3**|[reference](https://docs.microsoft.com/en-us/typography/opentype/spec/stat)|
| **PCLT**	PCL 5 data |*missing*|**0**|[reference](https://docs.microsoft.com/en-us/typography/opentype/spec/pclt)|
| **VDMX**	Vertical device metrics |*missing*|**1**|[reference](https://docs.microsoft.com/en-us/typography/opentype/spec/vdmx)|
| **vhea**	Vertical Metrics header |*missing*|**2**|[reference](https://docs.microsoft.com/en-us/typography/opentype/spec/vhea)|
| **vmtx**	Vertical Metrics |*missing*|**2**|[reference](https://docs.microsoft.com/en-us/typography/opentype/spec/vmtx)|

---

**<u>Outlines Tables</u>**

| description                         | status |
| ----------------------------------- | ------ |
| **glyf**	glyph data for TrueType | ✔️      |
| **CFF** 	Compact Font Format 1.0 | ✔️      |
| **CFF2**	Compact Font Format 2.0 | ✔️      |

---

**<u>Font Variations Tables</u>**

| description                                                  | status    | severity | reference                                                    |
| ------------------------------------------------------------ | --------- | -------- | ------------------------------------------------------------ |
| **avar**	Axis variations                                  | *missing* | **3**    | [reference](https://docs.microsoft.com/en-us/typography/opentype/spec/avar) |
| **cvar**	CVT variations (TrueType outlines only)          | *missing* | **3**    | [reference](https://docs.microsoft.com/en-us/typography/opentype/spec/cvar) |
| **fvar**	Font variations                                  | ✔️         |          |                                                              |
| **gvar**	Glyph variations (TrueType outlines only)        | *missing* | **3**    | [reference](https://docs.microsoft.com/en-us/typography/opentype/spec/gvar) |
| **HVAR**	Horizontal metrics variations                    | *missing* | **3**    | [reference](https://docs.microsoft.com/en-us/typography/opentype/spec/hvar) |
| **MVAR**	Metrics variations                               | *missing* | **3**    | [reference](https://docs.microsoft.com/en-us/typography/opentype/spec/mvar) |
| **STAT**	Style attributes (required for variable fonts, optional for non-variable fonts) | *missing* | **3**    | [reference](https://docs.microsoft.com/en-us/typography/opentype/spec/stat) |
| **VVAR**	Vertical metrics variations                      | *missing* | **3**    | [reference](https://docs.microsoft.com/en-us/typography/opentype/spec/vvar) |

---

**<u>Advanced Typographic Tables</u>**

| description                         | status           | severity | reference                                                    |
| ----------------------------------- | ---------------- | -------- | ------------------------------------------------------------ |
| **BASE**	Baseline data           | *missing*        | **4**    | [reference](https://docs.microsoft.com/en-us/typography/opentype/spec/base) |
| **GDEF**	Glyph definition data   | *missing*        | **4**    | [reference](https://docs.microsoft.com/en-us/typography/opentype/spec/gdef) |
| **GPOS**	Glyph positioning data  | *<u>partial</u>* | **4**    | [reference](https://docs.microsoft.com/en-us/typography/opentype/spec/gpos) |
| **GSUB**	Glyph substitution data | *<u>partial</u>* | **4**    | [reference](https://docs.microsoft.com/en-us/typography/opentype/spec/gsub) |
| **JSTF**	Justification data      | *missing*        | **3**    | [reference](https://docs.microsoft.com/en-us/typography/opentype/spec/jstf) |
| **MATH**	Math layout data        | *missing*        | **2**    | [reference](https://docs.microsoft.com/en-us/typography/opentype/spec/math) |

---

**<u>GSUB table status</u>**

| description                                | status    | severity | reference                                                    |
| ------------------------------------------ | --------- | -------- | ------------------------------------------------------------ |
| 1.1 Single Substitution format 1           | *missing* | **4**    | [reference](https://docs.microsoft.com/en-us/typography/opentype/spec/gsub#11-single-substitution-format-1) |
| 1.2 Single Substitution format 2           | ✔️         |          |                                                              |
| 2.1 Multiple Substitution format 1         | ✔️         |          |                                                              |
| 3.1 Alternate Substitution                 | missing   | **4**    | [reference](https://docs.microsoft.com/en-us/typography/opentype/spec/gsub#31-alternate-substitution-format-1) |
| 4.1 Ligature Substitution                  | *missing* | **4**    | [reference](https://docs.microsoft.com/en-us/typography/opentype/spec/gsub#41-ligature-substitution-format-1) |
| 5.1 Context Substitution Format 1          | *missing* | **4**    | [reference](https://docs.microsoft.com/en-us/typography/opentype/spec/gsub#51-context-substitution-format-1-simple-glyph-contexts) |
| 5.1 Context Substitution Format 2          | *missing* | **4**    | [reference](https://docs.microsoft.com/en-us/typography/opentype/spec/gsub#52-context-substitution-format-2-class-based-glyph-contexts) |
| 5.1 Context Substitution Format 3          | *missing* | **4**    | [reference](https://docs.microsoft.com/en-us/typography/opentype/spec/gsub#53-context-substitution-format-3-coverage-based-glyph-contexts) |
| 6.1 Chaining Context Substitution Format 1 | *missing* | **4**    | [reference](https://docs.microsoft.com/en-us/typography/opentype/spec/gsub#61-chaining-context-substitution-format-1-simple-glyph-contexts) |
| 6.2 Chaining Context Substitution Format 2 | *missing* | **4**    | [reference](https://docs.microsoft.com/en-us/typography/opentype/spec/gsub#62-chaining-context-substitution-format-2-class-based-glyph-contexts) |
| 6.3 Chaining Context Substitution Format 3 | ✔️         |          |                                                              |

---

<u>**GPOS table status**</u>

| description                      | status    | severity | reference                                                    |
| -------------------------------- | --------- | -------- | ------------------------------------------------------------ |
| 1	Single adjustment           | ✔️         |          |                                                              |
| 2	Pair adjustment             | ✔️         |          |                                                              |
| 3	Cursive attachment          | *missing* | **4**    | [reference](https://docs.microsoft.com/en-us/typography/opentype/spec/gpos#lookup-type-3-cursive-attachment-positioning-subtable) |
| 4	MarkToBase attachment       | *missing* | **4**    | [reference](https://docs.microsoft.com/en-us/typography/opentype/spec/gpos#lookup-type-4-mark-to-base-attachment-positioning-subtable) |
| 5	MarkToLigature attachment   | *missing* | **4**    | [reference](https://docs.microsoft.com/en-us/typography/opentype/spec/gpos#lookup-type-5-mark-to-ligature-attachment-positioning-subtable) |
| 6	MarkToMark attachment       | *missing* | **4**    | [reference](https://docs.microsoft.com/en-us/typography/opentype/spec/gpos#lookup-type-6-mark-to-mark-attachment-positioning-subtable) |
| 7	Context positioning         | *missing* | **4**    | [reference](https://docs.microsoft.com/en-us/typography/opentype/spec/gpos#lookup-type-7-contextual-positioning-subtables) |
| 8	Chained Context positioning | *missing* | **4**    | [reference](https://docs.microsoft.com/en-us/typography/opentype/spec/gpos#lookuptype-8-chaining-contextual-positioning-subtable) |
| 9	Extension positioning       | *missing* | **4**    | [reference](https://docs.microsoft.com/en-us/typography/opentype/spec/gpos#lookuptype-9-extension-positioning) |

---

**<u>Color Fonts & Bitmap Glyphs Tables</u>**

| description                                          | status    | severity | reference                                                    |
| ---------------------------------------------------- | --------- | -------- | ------------------------------------------------------------ |
| **COLR**	Color table                              | *missing* | **2** | [reference](https://docs.microsoft.com/en-us/typography/opentype/spec/colr) |
| **CPAL**	Color palette table                      | *missing* | **2** | [reference](https://docs.microsoft.com/en-us/typography/opentype/spec/cpal) |
| **CBDT**	Color bitmap data                        | *missing* | **2** | [reference](https://docs.microsoft.com/en-us/typography/opentype/spec/cbdt) |
| **CBLC**	Color bitmap location data               | *missing* | **2** | [reference](https://docs.microsoft.com/en-us/typography/opentype/spec/cblc) |
| **sbix**	Standard bitmap graphics                 | *missing* | **2** | [reference](https://docs.microsoft.com/en-us/typography/opentype/spec/sbix) |
| **SVG** 	The SVG (Scalable Vector Graphics) table | *missing* | **2** | [reference](https://docs.microsoft.com/en-us/typography/opentype/spec/svg) |
| **EBDT**	Embedded bitmap data |*missing*|**2**|[reference](https://docs.microsoft.com/en-us/typography/opentype/spec/ebdt)|
| **EBLC**	Embedded bitmap location data |*missing*|**2**|[reference](https://docs.microsoft.com/en-us/typography/opentype/spec/cblc)|
| **EBSC**	Embedded bitmap scaling data |*missing*|**2**|[reference](https://docs.microsoft.com/en-us/typography/opentype/spec/ebsc)|

---

- **(!) Hint,**

  Some tables like **GSUB** and **GPOS** requires multiple steps to be completed in a certain order, <u>for example</u> For a certain feature in GSUB table to be fully implemented, the library should support

  1. parsing ( reading data from font file )

  2. re-structuring data for future lookups

  3. the lookup logic that handles data retrieval to be used in rendering, ... etc.

  please make sure that you **don't repeat yourself** as some features has been already implemented and you can use it's existing logic or refactor it if you think you can do better.

---

**<u>Tooling</u>**

| description                                                  | status    | severity |
| ------------------------------------------------------------ | --------- | -------- |
| **Glyph Inspector**	Find a glyph by name, unicode or index | *missing* | **3**    |
| **Glyph Inspector**	Add sidebearing marks                 | *missing* | **2**    |
| **Glyph Inspector**	Fonts presets to easily inspect multiple languages | *missing* | **4**    |
| **Font Inspector**	Add formatting and syntax highlighting to GSUB and cmap table data | *missing* | **4**    |

---

**<u>Technical</u>**

* I recommend using **Typescript 3.0**

  using typescript will ensure that typings are synced with implementation unlike JSDocs. plus all the gains of using JSDocs. and it'll greatly improve code quality. what I propose is removing JSDocs type annotations while keeping description blocks and use typescript annotations instead.

---
I believe didn't cover everything, please feel free to **update** or **add** more items.

**Comment by yisibl** [comment]
Is there a plan to support Color Fonts?

**Comment by kontur** [comment]
woff2 decompression would be great (to be on the list).

**Comment by cmloegcmluin** [comment]
I'm very new to this world, so forgive me if I'm off-base. But I think I'm experiencing a problem using opentype.js because my font contains some GSUB lookup tables of type 7, extension substitution, which seem to be something like wrappers around lookup tables of other types (in my case, type 4, ligatures, which are otherwise working fine). Is there any plan to support GSUB type 7? Perhaps @bawarkamalqader was referring to this in the first post here, but @solomancode 's table cut off at 6.

**Comment by cmloegcmluin** [comment]
Ah, apologies. I was looking at a version of opentype.js that was a couple years old. It looks like extension substitution lookup tables may already be supported now. Thank you.

**Comment by Connum** [comment]
I added the Release 2.0.0 milestone, so we can re-evaluate the status quo.
We should especially
* update the nice overview in https://github.com/opentypejs/opentype.js/issues/374#issuecomment-485134088 and port it over to a more prominent place in either the Discussions or a new Wiki 
* have https://rawgit.com/unicode-org/text-rendering-tests/master/reports/OpenType.js.html updated with either a pre-release of 2.0.0 or at least 2.0.0 itself, so we can get a clearer perspective of what's next moving on from there

**Comment by ILOVEPIE** [comment]
> woff2 decompression would be great (to be on the list).

Woff2 decompression is out of scope for the library, you can use other libraries to do the decompression and then pass the results to opentype.js see https://github.com/opentypejs/opentype.js/issues/183#issuecomment-1147228025

**Comment by Connum** [comment]
I made a repository where I will store test reports from that tool for different versions and commits from the master branch. The reports are directly available to view on GitHub Pages here: https://connum.github.io/opentypejs-reports/
